doc: How to make a backup of Ceph with Velero (#6365)
rook/rook: Documentation/ceph-backup.md

---

> title: Back Up

Reviewer: English is very confusing, but I understand that the noun should be a single word and the verb is two words like you had it. Let's use the noun.
```suggestion
title: Backup
```

> weight: 2200

Reviewer: The weight indicates in what order the topics should be listed in the table of contents. This value will add it just before the "Ceph Tools" --> "Advanced Configuration".
```suggestion
weight: 11250
```

Author: Ah okey, I did not understand what value should be placed. Now I do, thank you!

indent: true
---

> # Back Up

Reviewer: ```suggestion
# Velero Backup
```

> Back Up of Ceph helps you recover your data when a disaster occurs.

Reviewer: ```suggestion
A backup of Ceph helps you recover your data when a disaster occurs.
```

## Requirements

The following requirements are needed:

1. Rook Ceph [PVC-based Cluster](ceph-cluster-crd.md#pvc-based-cluster).
2. [Velero](https://velero.io/) + [Restic](https://velero.io/docs/v1.5/restic/)

## Deploy of a Ceph PVC-based Cluster

### Deploy the operator

The first step is to deploy the Rook operator. Check that you are using the [example yaml files](https://github.com/rook/rook/blob/{{ branchName }}/cluster/examples/kubernetes/ceph) that correspond to your release of Rook. For more options, see the [examples documentation](ceph-examples.md).

```console
cd cluster/examples/kubernetes/ceph
kubectl create -f common.yaml
kubectl create -f operator.yaml

## verify the rook-ceph-operator is in the `Running` state before proceeding
kubectl -n rook-ceph get pod
```

You can also deploy the operator with the [Rook Helm Chart](helm-operator.md).

> ### Create a Rook Ceph PVC-based Cluster

Reviewer: Instead of describing how to setup the PVC-based cluster, we should have a link to another topic that says how to deploy it, then this topic just talks about the backup and restore steps. 

Since we don't have good documentation around PVC-based clusters, perhaps you can add a paragraph in [this topic](https://rook.github.io/docs/rook/v1.4/ceph-cluster-crd.html#pvc-based-cluster) to make it more clear how to deploy.

      #     storageClassName: io1
      #     volumeMode: Block
      #     accessModes:
      #       - ReadWriteOnce
      # dedicated block device to store bluestore wal (block.wal)
      # - metadata:
      #     name: wal
      #   spec:
      #     resources:
      #       requests:
      #         # Find the right size https://docs.ceph.com/docs/master/rados/configuration/bluestore-config-ref/#sizing
      #         storage: 5Gi
      #     # IMPORTANT: Change the storage class depending on your environment (e.g. local-storage, io1)
      #     storageClassName: io1
      #     volumeMode: Block
      #     accessModes:
      #       - ReadWriteOnce
      # Scheduler name for OSD pod placement
      # schedulerName: osd-scheduler
  resources:
  #  prepareosd:
  #    limits:
  #      cpu: "200m"
  #      memory: "200Mi"
  #   requests:
  #      cpu: "200m"
  #      memory: "200Mi"
  disruptionManagement:
    managePodBudgets: false
    osdMaintenanceTimeout: 30
    manageMachineDisruptionBudgets: false
    machineDisruptionBudgetNamespace: openshift-machine-api
```

### Create an Object Store to confirm that data is restored

> We are going to create an Object Store, create a bucket and add some items. So, on restore we can confirm that all data is restored.

Reviewer: Similar to the cluster section, let's not duplicate the instructions for setting up object storage. Instead, how about if we mention that these steps back up all of the block, file, and object storage?

Reviewer: Have you tried backing up block or file? Or just object? It would be good that this topic could cover all three. However, block and file might have a separate section since backing up rbd and cephfs volumes would be somewhat duplicated if the raw OSDs are backed up. So we might want to keep backup of block/file separate from the object storage backup.

Author: When I wrote that I have just tested object storage. However, this afternoon I have tried with filesystem. Tomorrow, I will test with block too.

The approach for filesystem is different.
As [documentation](https://rook.io/docs/rook/v1.4/ceph-filesystem.html) explains with filesystem you need to create a storage class that uses the filesystem. After that, your pods need to create a volumeclaim and mount it as a volume.

I got it working using the approach mentioned in this [link](https://velero.io/docs/v1.5/restic/#using-opt-in-pod-volume-backup).

export AWS_SECRET_ACCESS_KEY=`kubectl get secret -n rook-ceph ceph-delete-bucket -o 'jsonpath={.data.AWS_SECRET_ACCESS_KEY}' | base64 --decode;echo`

for i in {1..10};do aws s3 cp /etc/hosts s3://${BUCKET_NAME}/$i --endpoint-url http://${AWS_ENDPOINT};done

# List data
aws s3 ls s3://${BUCKET_NAME} --endpoint-url http://${AWS_ENDPOINT}
```

## Install and configure Velero

[Velero](https://velero.io/) is an open source tool to safely backup and restore, perform disaster recovery, and migrate Kubernetes cluster resources and persistent volumes.

For backing up and restoring Kubernetes volumes you need to use the [restic plugin](https://velero.io/docs/v1.5/restic/). You need to add `--use-restic` to velero install command.

Velero supports a lot of cloud providers. Take at look in the [supported providers](https://velero.io/docs/v1.5/supported-providers/) section of the doc for more information and examples. 

For example for [GCP](https://github.com/vmware-tanzu/velero-plugin-for-gcp#setup).

```console

# Configure Google Storage bucket and service account.

> BUCKET=<YOUR_BUCKET>

Reviewer: Is there Velero documentation that we could reference? We don't want to maintain documentation in Rook for Velero install instructions.

Author: Yes, the link above this code is the official documentation for Velero in GCP.

  --filter="displayName:Velero service account" \
  --format 'value(email)')

ROLE_PERMISSIONS=(
    compute.disks.get
    compute.disks.create
    compute.disks.createSnapshot
    compute.snapshots.get
    compute.snapshots.create
    compute.snapshots.useReadOnly
    compute.snapshots.delete
    compute.zones.get
)

gcloud iam roles create velero.server \
    --project $PROJECT_ID \
    --title "Velero Server" \
    --permissions "$(IFS=","; echo "${ROLE_PERMISSIONS[*]}")"

gcloud projects add-iam-policy-binding $PROJECT_ID \
    --member serviceAccount:$SERVICE_ACCOUNT_EMAIL \
    --role projects/$PROJECT_ID/roles/velero.server

gcloud iam service-accounts keys create credentials-velero \
    --iam-account $SERVICE_ACCOUNT_EMAIL

# Install velero
velero install --provider gcp --plugins velero/velero-plugin-for-gcp:v1.1.0 --bucket $BUCKET --secret-file ./credentials-velero --use-restic

```

## Create Back Up

We need to save Volumes, Volumes Claims, ConfigMaps, Secrets and Services.

```console
velero backup create <BackupName> --include-namespaces rook-ceph --include-resources pv,pvc,services,configmaps,secrets --default-volumes-to-restic
```

> **IMPORTANT:** make sure to back up the bucket's configmaps and secrets that are located in a different namespace.

Reviewer: Which different namespace? The application namespace?
```suggestion
**IMPORTANT:** In addition to backing up the resources in the `rook-ceph` namespace, make sure to back up the bucket's configmaps and secrets that are located in the application namespace.
```

Author: When you execute for example `object-bucket-claim-delete.yaml`. If you set  `namespace`, configmaps and secrets will go to other namespace

```

We end up with:
```yaml
apiVersion: v1
kind: Service
metadata:
  labels:
    app: rook-ceph-mon
    ceph_daemon_id: a
    ceph_daemon_type: mon
    mon: a
    mon_cluster: rook-ceph
    pvc_name: rook-ceph-mon-a
    rook_cluster: rook-ceph
  name: rook-ceph-mon-a
  namespace: rook-ceph
spec:
  clusterIP: 10.4.1.210
  ports:
  - name: tcp-msgr1
    port: 6789
    protocol: TCP
    targetPort: 6789
  - name: tcp-msgr2
    port: 3300
    protocol: TCP
    targetPort: 3300
  selector:
    app: rook-ceph-mon
    ceph_daemon_id: a
    mon: a
    mon_cluster: rook-ceph
    rook_cluster: rook-ceph
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```

## Restore Back Up

Execute the following command to restore the backup.

```console
velero restore create --from-backup <BackupName>
```

Delete the `MON` services backed up by Velero and restore the old ones.

```console
# Delete MONs
kubectl delete svc -n rook-ceph rook-ceph-mon-a
kubectl delete svc -n rook-ceph rook-ceph-mon-b
kubectl delete svc -n rook-ceph rook-ceph-mon-c

# Restore the new ones
kubectl apply -f mon-service-a.yaml
kubectl apply -f mon-service-b.yaml
kubectl apply -f mon-service-c.yaml
```

> After that we need to create cluster.

Reviewer: ```suggestion
Now we need to create cluster with the same settings that were used in the original cluster. In particular, the cluster-on-pvc.yaml settings should be identical or else your cluster may not be restored as expected.
```

```console
kubectl apply -f common.yaml
kubectl apply -f operator.yaml
kubectl apply -f cluster-on-pvc.yaml
```

> Owner reference will be attached to `MON` services.

Reviewer: Why mention the owner reference?

Author: When we export the service for the first time we get a yaml as follows:

```yaml
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: "2020-10-04T19:22:31Z"
  labels:
    app: rook-ceph-mon
    ceph_daemon_id: a
    ceph_daemon_type: mon
    mon: a
    mon_cluster: rook-ceph
    pvc_name: rook-ceph-mon-a
    rook_cluster: rook-ceph
  name: rook-ceph-mon-a
  namespace: rook-ceph
  ownerReferences:
  - apiVersion: ceph.rook.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: CephCluster
    name: rook-ceph
    uid: 7561619e-bd35-4649-8af7-a7a0dd7c11b4
  resourceVersion: "3046"
  selfLink: /api/v1/namespaces/rook-ceph/services/rook-ceph-mon-a
  uid: 77e161de-d9ed-41eb-b9c2-386393d21369
spec:
  clusterIP: 10.4.1.210
  ports:
  - name: tcp-msgr1
    port: 6789
    protocol: TCP
    targetPort: 6789
  - name: tcp-msgr2
    port: 3300
    protocol: TCP
    targetPort: 3300
  selector:
    app: rook-ceph-mon
    ceph_daemon_id: a
    mon: a
    mon_cluster: rook-ceph
    rook_cluster: rook-ceph
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```

I am explaining that you need to delete `metadata.ownerReferences`. However, they will get added by the operator in the recreation
